### PR TITLE
add non-linear fit notebooks

### PR DESCRIPTION
**Pull request by friedrichknuth**
- Add `plotting.py` and `io.py` libraries. 
  - `io.py` contains functions to read/stack geotifs with rioxarray and optionally write to netcdf. 
  - Reprojects to match input reference DEM as needed.

- Add `temporal_fitting_non_linear.ipynb` notebook for exploratory non-linear GPR fitting.
- Add `temporal_fitting_non_linear_centerline_xr.ipynb` notebook for time series stacking with rioxarray and point selection at geospatial coordinates. 
  - Includes function to pull in `OGGM` glacier center flowlines and use to sample time series at geospatial coordinates.
- Update time series gallery plotting function to plot point coordinates as subplot titles and include y axis.
- Miscellaneous updates for improved masking, plotting, wrangling

### CODE REVIEW

**Comment by ShashankBice** [comment]
Hi Friedrich, the updates look great !
One suggestion on the non-linear temporal fit notebook. Right now you have axis-labels and then plot the timeseries for the points with their spatial coordinates, but it could still be confusing. One thing which we might try is to use different markers or different color of markers, so that the users can correlate the spatial location on the point on the mapview plot [Cell 49] and them compare it with the elevation timeseries plot [cell 16] more easily.
The GPR plots are great with the y-axis labels now :)
The second notebook on the linear fits along centerline plots is great, and some of the time-series plot depict that they may warrant different shorter linear trends fit to them (a persistent decline in elevation followed by a rise (e.g., 615981,3955878)).

**Comment by friedrichknuth** [comment]
Thanks @ShashankBice ! Different colors for the points are a great idea. Maybe we can add that as an enhancement in a future PR? In the mean time, the point coordinate labels should help see which time series subplots belong to which point. Please indicate if you approve merging the PR as is, for now.

**Review by ShashankBice — APPROVED** [review]

**Comment by dshean** [comment]
Looks great!
- I had same thought on point colors/markers - would make it much easier to tie the time series plots back to the map.  Or add point numbers to map and time series, or add coord labels on the map.
    - Not sure if you ever tried the old `stack_view.py` tool in imview.  It likely needs some updates, but it allowed for quick interactive point selection and plotting.  Had to some solutions for the color cycle handling there.
- Linear distance from the first point in the profile could also help with interpretation.
- The dh/dt map shows coherent signals related to "surges" near the terminus and along lower tributaries, some with positive dh/dt, some with negative dh/dt.  Could be useful to review dh/dt map and manually place a few points on each of those areas to consider the temporal evolution.

**Comment by friedrichknuth** [comment]
Sounds good! I will check out the `stack_view.py` tool. 

Regarding color cycling, I've lately just been doing something like `color='C'+str(idx)` as I loop through things, which cycles through the default matplotlib colors (e.g. `C0` is the default mpl blue, though I'm guessing there is a limit to `Cn` somewhere. Otherwise, been using the XKCD cycles :) 

```
matplotlib._color_data as mcd
cycle = list(mcd.XKCD_COLORS.values())
...color=cycle[idx]
```

Will look into grabbing the dh/dt map to identify some more points of interest. Can add something to plot with linear distance, when dealing with line segments (vs randomly selected points).

@dshean If this looks ok to merge for now, can you indicate by approving the PR?

**Comment by friedrichknuth** [comment]
Latest commit should resolve #8

Updates in `main_experiment2_temporal.py` Should run faster now with using `pool.map` instead of `pool.imap`.

Was having some issues with `np.ma.count(dem_stack.data)`. For some reason I could not decipher this treats all values as valid, so masking out pixels with < `n_thresh` observations does not work correctly and causes the linear regression to fail where all values are invalid, for example. Note that `dem_stack.data` is the masked array returned by `dem_stack = gu.spatial_tools.stack_rasters()` Used rioxarray to stack for now. 

More details in the `main_experiment2_temporal.py` and `temporal_fitting_analysis.ipynb` notebook.

**Comment by adehecq** [comment]
Will merge and go from there. Thanks !

File: ragmac_xdem/io.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    if len(datetimes_list) == len(geotif_files_list):
+        pass
+    else:
+        print('length of datetimes does not latch length of GeoTIF file list')
+        print('datetimes:',len(datetimes_list))
+        print('geotifs:',len(geotif_files_list))
+        return None
+
+    ## Choose resampling method. Defaults to bilinear.
+    if isinstance(resampling, type(Resampling.bilinear)):
```
Comment (adehecq) [inline ragmac_xdem/io.py:107]: You could use one of geoutils functions here: https://github.com/GlacioHack/GeoUtils/blob/dfc4add19f163f7f5e47ffd0e175953b0175513d/geoutils/misc.py#L157

File: ragmac_xdem/temporal.py [hunk 2]
```diff
@@ -310,3 +319,194 @@ def ma_linreg(
...
+    
+    return slope, intercept
+
+def linreg_predict(args):
+    slope, x, intercept = args
+    prediction = slope * x + intercept
+    return prediction
+
+def linreg_predict_parallel(slope,X,intercept):
+    pool = mp.Pool(processes=psutil.cpu_count(logical=True))
```
Comment (adehecq) [inline ragmac_xdem/temporal.py:392]: would be good to make it an optional argument, and pass the option in the main scripts.

File: scripts/main_experiment2_temporal.py [hunk 3]
```diff
@@ -101,42 +101,121 @@
...
+
+        now = datetime.now()
+        dt = now-start
+        print('Elapsed time:', str(dt).split(".")[0])
+
+
+        print('\n### Compute residuals ###')
+        prediction = temporal.linreg_predict_parallel(slope,X_train,intercept)
+        residuals  = ma_stack - prediction
+        residuals[:,valid_mask_2D]  = residuals[:,valid_mask_2D]
```
Comment (adehecq) [inline scripts/main_experiment2_temporal.py:136]: You simply wrote `a=a`.

File: scripts/main_experiment2_temporal.py [hunk 4]
```diff
@@ -101,42 +101,121 @@
...
+        
+        now = datetime.now()
+        dt = now-start
+        print('Elapsed time:', str(dt).split(".")[0])
+        
+        print('\n### Compute MAD ###')
+        detrended_std = temporal.mad(residuals, axis=0)
```
Comment (adehecq) [inline scripts/main_experiment2_temporal.py:143]: This step took about 10 min on my laptop ! What is it so slow?

**Pull request merged**